### PR TITLE
Add api for password change

### PR DESCRIPTION
**Pull request by k9845**
## Added api for password-change
- api `/api/v1/users//me/change-password/`
Post request with `old_password` and 'new_password`
## Added information in the password_change template
`Time` `Location` `Browser`

## Related Issue:
- #770 
Addresses
- the-deep/client#1682
@samshara

### CODE REVIEW

File: apps/user/tests/test_apis.py [hunk 1]
```diff
@@ -216,3 +220,49 @@ def test_user_preference_feature_avialable_for_all(self):
...
+        assert user.check_password(data['new_password'])
+
+        # now try with posting  diferent `new_password` that doesnot follow django password validation
+        data = {
+            "old_password": new_pass,  # since password is already changed in the database level
+            "new_password": "nepa"
+        }
+        self.authenticate(user)
+        response = self.client.post(url, data)
+        print(response.content)
```
Comment (thenav56) [inline apps/user/tests/test_apis.py]: Let's make sure to https://github.com/the-deep/server/blob/develop/.github/pull_request_template.md#make-sure-there-are-no

File: apps/user/tests/test_apis.py [hunk 2]
```diff
@@ -216,3 +220,49 @@ def test_user_preference_feature_avialable_for_all(self):
...
+        self.assert_400(response)
+
+        # now try with posting different `old_password`
+        data = {
+            "old_password": "hahahmeme",
+            "new_password": "nepa"
+        }
+        self.authenticate(user)
+        response = self.client.post(url, data)
+        print(response.content)
```
Comment (thenav56) [inline apps/user/tests/test_apis.py]: Let's make sure to https://github.com/the-deep/server/blob/develop/.github/pull_request_template.md#make-sure-there-are-no

File: apps/user/views.py [hunk 3]
```diff
@@ -139,6 +140,21 @@ def post(self, request, version=None):
...
+    def post(self, request, version=None):
+        serializer = PasswordChangeSerializer(
+            data=request.data,
+            context={
+                'request': request
+            }
+        )
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        return response.Response(serializer.data, status=status.HTTP_201_CREATED)
```
Comment (thenav56) [inline apps/user/views.py]: Let' use 200 instead of 201.

File: apps/user/serializers.py [hunk 4]
```diff
@@ -222,3 +227,28 @@ class EntryCommentUserSerializer(serializers.ModelSerializer):
     class Meta:
         model = User
         fields = ('id', 'name', 'email', 'organization', 'display_picture_url',)
+
+
+class PasswordChangeSerializer(serializers.ModelSerializer):
```
Comment (thenav56) [inline apps/user/serializers.py]: serializers.ModelSerializer is not required. Let's use serializers.Serializer.

File: apps/user/serializers.py [hunk 5]
```diff
@@ -222,3 +227,28 @@ class EntryCommentUserSerializer(serializers.ModelSerializer):
...
+    old_password = serializers.CharField(required=True, write_only=True)
+    new_password = serializers.CharField(required=True, write_only=True)
+
+    class Meta:
+        model = User
+        fields = ['old_password', 'new_password']
```
Comment (thenav56) [inline apps/user/serializers.py]: Not required.

File: apps/user/serializers.py [hunk 6]
```diff
@@ -222,3 +227,28 @@ class EntryCommentUserSerializer(serializers.ModelSerializer):
...
+
+    def validate_new_password(self, password):
+        validate_password(password)
+        return password
+
+    def save(self, **kwargs):
+        user = self.context['request'].user
+        user.set_password(self.validated_data['new_password'])
+        user.save()
+        send_password_change(user)
```
Comment (thenav56) [inline apps/user/serializers.py]: Let's do 3 things here.
- Use transaction.on_commit
- change send_password_change -> send_password_changed_notification
- Make send_password_changed_notification celery task and use that.

```python3
@shared_task
def send_password_changed_notification(user_id):
    .....
....
transaction.on_commit(
    lambda: send_password_changed_notification.delay(user.id)
)
```

File: apps/user/models.py [hunk 7]
```diff
@@ -27,6 +27,7 @@ class Profile(models.Model):
     # Email Conditions
     E_ACCOUNT_ACTIVATION = 'account_activation'
     E_PASSWORD_RESET = 'password_reset'
+    E_PASSWORD_CHANGE = 'password_change'
```
Comment (thenav56) [inline apps/user/models.py]: `E_PASSWORD_CHANGED = 'password_changed'`

File: apps/templates/password_change/password_change_email.html [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{% block container %}
+    <p>
+        Your password has been changed sucessfully.
```
Comment (thenav56) [inline apps/templates/password_change/password_change_email.html]: Is this okay? @AdityaKhatri

File: apps/templates/password_change/password_change_subject.txt [hunk 9]
```diff
@@ -0,0 +1,4 @@
+{% load i18n %}
+{% autoescape off %}
+DEEP Password Change
```
Comment (thenav56) [inline apps/templates/password_change/password_change_subject.txt]:
```suggestion
DEEP Password Changed
```

File: apps/user/views.py [hunk 10]
```diff
@@ -139,6 +140,21 @@ def post(self, request, version=None):
             serializer.data, status=status.HTTP_201_CREATED)
 
 
+class PasswordChangeView(views.APIView):
+    permission_classes = [permissions.IsAuthenticated]
```
Comment (thenav56) [inline apps/user/views.py]: Let's use action (detail=False, path='me/change_password') route in UserViewSet

File: deep/urls.py [hunk 11]
```diff
@@ -428,6 +429,9 @@ def get_api_path(path):
             )
         ),
         name='password_reset_confirm'),
+    # Password Change View
+    url(get_api_path(r'password-change/$'),
+        PasswordChangeView.as_view()),
```
Comment (thenav56) [inline deep/urls.py]: No need here, let's use the UserViewSet action route.

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/templates/password_change/password_change_subject.txt [hunk 12]
```diff
@@ -0,0 +1,4 @@
+{% load i18n %}
```
Comment (thenav56) [inline apps/templates/password_change/password_change_subject.txt]: Make sure to change the filename as well.

File: apps/user/serializers.py (hunk 5 above)
Comment (thenav56) [reply apps/user/serializers.py]: This is also not required. `fields = [..`

File: apps/templates/password_change/password_change_subject.txt (hunk 9 above)
Comment (thenav56) [reply apps/templates/password_change/password_change_subject.txt]: Also change the filename.

File: apps/user/tests/test_apis.py [hunk 13]
```diff
@@ -13,6 +13,10 @@
 
 
 class UserApiTests(TestCase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.user_password = 'joHnDave!@#123'
+
```
Comment (thenav56) [inline apps/user/tests/test_apis.py]: Do we need this? or define this here?

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/user/tests/test_apis.py (hunk 13 above)
Comment (k9845) [reply apps/user/tests/test_apis.py]: not..changed in `e640daa`

File: apps/user/views.py [hunk 14]
```diff
@@ -129,6 +130,24 @@ def get_notifications(self, request, pk=None, version=None):
...
+    def change_password(self, request, pk=None, version=None):
+        serializer = PasswordChangeSerializer(
+            data=request.data,
+            context={
+                'request': request
+            }
+        )
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        return response.Response(serializer.data, status=status.HTTP_200_OK)
```
Comment (thenav56) [inline apps/user/views.py]: Let's remove `serializer.data` from here. Don't think we need to send the password back.

File: apps/user/utils.py [hunk 15]
```diff
@@ -160,3 +161,14 @@ def send_project_join_request_emails(join_request_id):
...
+
+
+@shared_task
+def send_password_changed_notification(user_id):
+    user = User.objects.get(pk=user_id)
+    send_mail_to_user(
+        user, email_type=Profile.E_PASSWORD_CHANGED,
+        context={},
+        subject_template_name='password_change/subject.txt',
+        email_template_name='password_change/email.html',
```
Comment (thenav56) [inline apps/user/utils.py]:
```suggestion
        subject_template_name='password_changed/subject.txt',
        email_template_name='password_changed/email.html',
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/templates/password_change/email.html [hunk 16]
```diff
@@ -0,0 +1,5 @@
+{% block container %}
+    <p>
+        Your password has been changed sucessfully.
```
Comment (AdityaKhatri) [inline apps/templates/password_change/email.html]: @thenav56  I've replied here.
```suggestion
        Your password has been changed successfully.
```

File: apps/user/permissions.py [hunk 17]
```diff
@@ -2,15 +2,16 @@
 
 
 class UserPermission(permissions.BasePermission):
-    def has_object_permission(self, request, view, obj):
-        if request.method in permissions.SAFE_METHODS:
-            return True
-        return obj == request.user
-
+    def _is_authenticated(self, rq):
+        return rq.user and rq.user.is_authenticated
```
Comment (annshress) [inline apps/user/permissions.py]: Not the most important thing.

`request.user` is always going to be either `User` instance or `contrib.auth.AnonymousUser` instance.

```suggestion
    def _is_authenticated(self, rq):
        return rq.user.is_authenticated
```

**Review by annshress — APPROVED** [review]

**Review by annshress — APPROVED** [review]

**Review by thenav56 — APPROVED** [review]
All good.

@AdityaKhatri Make sure to review this file apps/templates/password_changed/email.html

File: apps/templates/password_changed/email.html [hunk 18]
```diff
@@ -0,0 +1,14 @@
+
+{% block container %}
+    <p>
+        Your password has been changed successfully.
+        Time:{{time}}
+        Location:{{location}}
+        Device:{{device}}
+        <a class="button-link"
```
Comment (AdityaKhatri) [inline apps/templates/password_changed/email.html]:
```suggestion
        Device:{{device}}
        If you are aware of this change, you can disregard this email.
        If this was not triggered by you, click the button below to reset your password.
        <a class="button-link"
```

File: apps/templates/password_changed/email.html [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+        href="{{ protocol }}://{{ domain }}{% url 'user-change_password' version='v2' %}"
+        >
+        I didn't do this--change again
```
Comment (AdityaKhatri) [inline apps/templates/password_changed/email.html]:
```suggestion
Reset Password
```

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

File: apps/templates/password_changed/email.html (hunk 18 above)
Comment (k9845) [reply apps/templates/password_changed/email.html]: Done

File: apps/templates/password_changed/email.html (hunk 19 above)
Comment (k9845) [reply apps/templates/password_changed/email.html]: Done

File: apps/templates/password_changed/email.html [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+{% block container %}
+    <p>
+        Your password has been changed successfully.
+        Time:{{time}}
+        Location:{{location}}
+        Device:{{device}}
+        If you are aware of this change, you can disregard this email.
+        If this was not triggered by you, click the button below to reset your password.
+        <a class="button-link"
+        href="{{ protocol }}://{{ domain }}{% url 'user-change_password' version='v2' %}"
```
Comment (thenav56) [inline apps/templates/password_changed/email.html]: `{% url 'user-change_password' version='v2' %}` where does this point to?

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/templates/password_changed/email.html (hunk 20 above)
Comment (k9845) [reply apps/templates/password_changed/email.html]: Point to `/api/v2/users//me/change-password/` as suggested earlier

Comment (k9845) [reply apps/templates/password_changed/email.html]: Changed that in  commit `1f3833c`

**Review by thenav56 — APPROVED** [review]

**Pull request merged**